Docs: User guide overhaul, part 2 (#65474)
ansible/ansible: docs/docsite/rst/user_guide/playbooks_filters.rst

how to create custom Ansible filters as plugins, though we generally welcome new filters into the core code so everyone can use them.

Templating happens on the Ansible controller, **not** on the target host, so filters execute on the controller and manipulate data locally.

.. contents::
   :local:

Handling undefined variables
============================

Filters can help you manage missing or undefined variables by providing defaults or allowing null values. If you configure Ansible to ignore most undefined variables, you can mark some variables as requiring values with the ``mandatory`` filter.

.. _defaulting_undefined_variables:

Providing default values
------------------------

You can provide default values for variables directly in your templates using the Jinja2 'default' filter. This is often a better approach than failing if a variable is not defined::

    {{ some_variable | default(5) }}

> In the above example, if the variable 'some_variable' is not defined, Ansibie uses the default value 5, rather than raising an "undefined variable" error and failing. If you are working within a role, you can also add a ``defaults/main.yml`` to define the default values for variables in your role.

Reviewer: ```suggestion
In the above example, if the variable 'some_variable' is not defined, Ansible uses the default value 5, rather than raising an "undefined variable" error and failing. If you are working within a role, you can also add a ``defaults/main.yml`` to define the default values for variables in your role.
```

Beginning in version 2.8, attempting to access an attribute of an Undefined value in Jinja will return another Undefined value, rather than throwing an error immediately. This means that you can now simply use
a default with a value in a nested data structure (i.e :code:`{{ foo.bar.baz | default('DEFAULT') }}`) when you do not know if the intermediate values are defined.

If you want to use the default value when variables evaluate to false or an empty string you have to set the second parameter to ``true``::

    {{ lookup('env', 'MY_USER') | default('admin', true) }}

.. _omitting_undefined_variables:

Omit: allowing null values
--------------------------

> If you want to allow null values for a variable, use the special ``omit`` variable as a default value::

Reviewer 2: I think the meaning here has changed.

It means that the argument is skipped if null

Author: I was working from the example. If the default of `omit` wasn't present, the example would return an undefined variable error, right?

Reviewer 2: ```suggestion
It is possible to use the default filter to skip module parameters using the special `omit` variable::
```

Author: Hmmm. After thinking about this for a while, maybe the right phrasing is:
"If you want to make a module parameter optional in a variable, use the special 'omit' variable as the default value. With this setup, Ansible will use the parameter if you supply it and skip it if you leave it out."
Does this same approach work for variables other than module parameters?

Author: Per @s-hertel, that approach will work for any variable. Updating the page now.

    {{ tags | items2dict(key_name='key', value_name='value') }}

Discovering the data type
-------------------------

.. versionadded:: 2.3

If you are unsure of the underlying Python type of a variable, you can use the ``type_debug`` filter to display it. This is useful in debugging when you need a particular type of variable::

    {{ myvar | type_debug }}

> Setting the data type

Reviewer 2: Setting or forcing?

---------------------

> You can cast values as certain types. This is especially useful for ensuring that Ansible knows when to treat a value like "True" as a boolean and when to treat it as a string::

Reviewer: ```suggestion
You can cast values as certain types. This is especially useful for ensuring that Ansible knows when to treat a value like "True" as a Boolean and when to treat it as a string::
```

Reviewer: TIL: Boolean is supposed to be capitalized... who knew?  :-)

Author: Huh. I bet we have a lot of places where that's wrong.

   - debug:
       msg: test
     when: some_string_value | bool

.. versionadded:: 1.6

.. _filters_for_formatting_data:

Controlling data formats: YAML and JSON
=======================================

The following filters will take a data structure in a template and manipulate it or switch it from or to JSON or YAML format. These are occasionally useful for debugging::

    {{ some_variable | to_json }}
    {{ some_variable | to_yaml }}

For human readable output, you can use::

    {{ some_variable | to_nice_json }}
    {{ some_variable | to_nice_yaml }}

> You can change the indentation of either format (new in version 2.2)::

Reviewer 2: I think 2.2 is too old to be notable 

```suggestion
You can change the indentation of either format::
```

    {{ some_variable | to_nice_json(indent=2) }}
    {{ some_variable | to_nice_yaml(indent=8) }}

The ``to_yaml`` and ``to_nice_yaml`` filters use `PyYAML library`_ which has a default 80 symbol string length limit. That causes unexpected line break after 80th symbol (if there is a space after 80th symbol)
To avoid such behavior and generate long lines, use the ``width`` option::

    {{ some_variable | to_yaml(indent=8, width=1337) }}
    {{ some_variable | to_nice_yaml(indent=8, width=1337) }}

> While it would be nicer to use a construction like ``float("inf")`` instead of a hardcoded number, unfortunately the filter doesn't support proxying Python functions.

Reviewer 2: I wonder if this should be a `note`?

Author: Yeah, it's a bit awkward. I'll try incorporating the information above the example.

Note that it also supports passing through other YAML parameters. Full list can be found in `PyYAML documentation`_.

If you are reading in some already formatted data::

    {{ some_variable | from_json }}
    {{ some_variable | from_yaml }}

for example::

  tasks:
    - shell: cat /some/path/to/file.json
      register: result

    - set_fact:
        myvar: "{{ result.stdout | from_json }}"

.. versionadded:: 2.7

> To parse multi-document yaml strings, the ``from_yaml_all`` filter is provided.

Reviewer: ```suggestion
To parse multi-document YAML strings, the ``from_yaml_all`` filter is provided.
```

> The ``from_yaml_all`` filter will return a generator of parsed yaml documents.

Reviewer: ```suggestion
The ``from_yaml_all`` filter will return a generator of parsed YAML documents.
```

This behavior does not depend on the value of the `hash_behavior` setting in `ansible.cfg`.

.. _extract_filter:

Selecting values from arrays or hashtables
-------------------------------------------

.. versionadded:: 2.1

The `extract` filter is used to map from a list of indices to a list of
values from a container (hash or array)::

    {{ [0,2] | map('extract', ['x','y','z']) | list }}
    {{ ['x','y'] | map('extract', {'x': 42, 'y': 31}) | list }}

The results of the above expressions would be::

    ['x', 'z']
    [42, 31]

The filter can take another argument::

    {{ groups['x'] | map('extract', hostvars, 'ec2_ip_address') | list }}

This takes the list of hosts in group 'x', looks them up in `hostvars`,
and then looks up the `ec2_ip_address` of the result. The final result
is a list of IP addresses for the hosts in group 'x'.

The third argument to the filter can also be a list, for a recursive
lookup inside the container::

    {{ ['a'] | map('extract', b, ['x','y']) | list }}

This would return a list containing the value of `b['a']['x']['y']`.

Combining lists
---------------

This set of filters returns a list of combined lists.

> .. versionadded:: 2.3

Reviewer 2: too old to be notable
```suggestion
```

permutations
^^^^^^^^^^^^
To get permutations of a list::

    - name: give me largest permutations (order matters)
      debug:
        msg: "{{ [1,2,3,4,5] | permutations | list }}"

    - name: give me permutations of sets of three
      debug:
        msg: "{{ [1,2,3,4,5] | permutations(3) | list }}"

combinations
^^^^^^^^^^^^
Combinations always require a set size::

    - name: give me combinations for sets of two
      debug:
        msg: "{{ [1,2,3,4,5] | combinations(2) | list }}"

Also see the :ref:`zip_filter`

products
^^^^^^^^
The product filter returns the `cartesian product <https://docs.python.org/3/library/itertools.html#itertools.product>`_ of the input iterables.

This is roughly equivalent to nested for-loops in a generator expression.

For example::

  - name: generate multiple hostnames
    debug:
      msg: "{{ ['foo', 'bar'] | product(['com']) | map('join', '.') | join(',') }}"

This would result in::

    { "msg": "foo.com,bar.com" }

.. json_query_filter:

Selecting JSON data: JSON queries
---------------------------------

> .. versionadded:: 2.2

Reviewer 2: ```suggestion
```

    - name: "Display all ports from cluster1"
      debug:
        var: item
      loop: "{{ domain_definition | json_query('domain.server[?cluster==''cluster1''].port') }}"

.. note:: Escaping single quotes within single quotes in YAML is done by doubling the single quote.

In this example, we get a hash map with all ports and names of a cluster::

    - name: "Display all server ports and names from cluster1"
      debug:
        var: item
      loop: "{{ domain_definition | json_query(server_name_cluster1_query) }}"
      vars:
        server_name_cluster1_query: "domain.server[?cluster=='cluster2'].{name: name, port: port}"

Randomizing data
================

When you need a randomly generated value, use one of these filters.

.. _random_mac_filter:

Random MAC addresses
--------------------

.. versionadded:: 2.6

This filter can be used to generate a random MAC address from a string prefix.

To get a random MAC address from a string prefix starting with '52:54:00'::

    "{{ '52:54:00' | random_mac }}"
    # => '52:54:00:ef:1c:03'

Note that if anything is wrong with the prefix string, the filter will issue an error.

> 

Reviewer 2: ```suggestion

 .. versionadded:: 2.9
```

As of Ansible version 2.9, it's also possible to initialize the random number generator from a seed. This way, you can create random-but-idempotent MAC addresses::

    "{{ '52:54:00' | random_mac(seed=inventory_hostname) }}"

.. _random_filter:

Random items or numbers
-----------------------

> .. versionadded:: 1.6

Reviewer 2: ```suggestion
```

> This filter can be used similar to the default jinja2 random filter (returning a random item from a sequence of

Reviewer: ```suggestion
This filter can be used similar to the default Jinja2 random filter (returning a random item from a sequence of
```

Reviewer: obviously, the further along I get, the more nitpicky... another TIL - Jinja2 is initial Cap according to their documentation...

Author: Thanks for finding the nits!

items), but can also generate a random number based on a range.

To get a random item from a list::

    "{{ ['a','b','c'] | random }}"
    # => 'c'

To get a random number between 0 and a specified number::

    "{{ 60 | random }} * * * * root /script/from/cron"
    # => '21 * * * * root /script/from/cron'

Get a random number from 0 to 100 but in steps of 10::

    {{ 101 | random(step=10) }}
    # => 70

Get a random number from 1 to 100 but in steps of 10::

    {{ 101 | random(1, 10) }}
    # => 31
    {{ 101 | random(start=1, step=10) }}
    # => 51

> As of Ansible version 2.3, it's also possible to initialize the random number generator from a seed. This way, you can create random-but-idempotent numbers::

Reviewer 2: ```suggestion
It's also possible to initialize the random number generator from a seed. This way, you can create random-but-idempotent numbers::
```

    "{{ 60 | random(seed=inventory_hostname) }} * * * * root /script/from/cron"

Shuffling a list
----------------

> .. versionadded:: 1.8

Reviewer 2: ```suggestion
```

This filter will randomize an existing list, giving a different order every invocation.

To get a random list from an existing  list::

    {{ ['a','b','c'] | shuffle }}
    # => ['c','a','b']
    {{ ['a','b','c'] | shuffle }}
    # => ['b','c','a']

> As of Ansible version 2.3, it's also possible to shuffle a list idempotent. All you need is a seed.::

Reviewer 2: ```suggestion
It's also possible to shuffle a list idempotent. All you need is a seed.::
```

    {{ ['a','b','c'] | shuffle(seed=inventory_hostname) }}
    # => ['b','a','c']

> note that when used with a non 'listable' item it is a noop, otherwise it always returns a list

Reviewer 2: define noop?

Author: Ah, good catch. I changed the wording around to express the meaning in plain language.

.. _list_filters:

List filters
============

These filters all operate on list variables.

> .. versionadded:: 1.8

Reviewer 2: ```suggestion
```

    {{ 'secretpassword' | password_hash('sha256', 'mysecretsalt', rounds=10000) }}

.. _other_useful_filters:

Text filters
============

These filters work with strings and text.

.. _comment_filter:

Adding comments to files
------------------------

> .. versionadded:: 2.0

Reviewer 2: ```suggestion
```

.. versionadded:: 1.9

To make use of one attribute from each item in a list of complex variables, use the "map" filter (see the `Jinja2 map() docs`_ for more)::

    # get a comma-separated list of the mount points (e.g. "/,/mnt/stuff") on a host
    {{ ansible_mounts | map(attribute='mount') | join(',') }}

Date and time filters
=====================

> To get a date object from a string use the `to_datetime` filter, (new in version in 2.2)::

Reviewer 2: ```suggestion
To get a date object from a string use the `to_datetime` filter::
```
